### ISSUE TITLE
PatchDataset / PatchLoader for torch & JAX

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/geopatcher#23]
**Labels**: `enhancement`, `ml`, `epic:ml`

### User story

As an ML user, I want `patcher.as_torch_dataset(field)` / `patcher.as_jax_loader(field)` so I don't write the glue.

### Motivation

Patch-as-batch *is* the dominant ML use case. Every user re-implements this in ~20 lines that are subtly wrong (memory pinning, shuffling semantics, distributed sampler partitioning). Ship it once, correctly.

### Mathematics

N/A.

### API proposal

```python
# Torch
from geopatcher.torch import PatchDataset
ds = PatchDataset(field, patcher, transform=norm_transform)
loader = torch.utils.data.DataLoader(ds, batch_size=32, num_workers=4, pin_memory=True)

# JAX
from geopatcher.jax import PatchLoader
loader = PatchLoader(field, patcher, batch_size=32, shuffle=True, seed=0, prefetch=2)
```

### Example usage

```python
# Inference over a Sentinel-2 tile with a torch model
import torch
from geopatcher.torch import PatchDataset

ds = PatchDataset(field, patcher)
loader = torch.utils.data.DataLoader(ds, batch_size=32, num_workers=4)

outs = []
for batch in loader:
    with torch.no_grad():
        pred = model(batch["data"].cuda())
    for i, anchor in enumerate(batch["anchor"]):
        outs.append(gp.Patch(data=pred[i].cpu().numpy(), anchor=anchor))

stitched = patcher.merge(outs, field.domain)
```

### Implementation plan

- [ ] `geopatcher/torch/__init__.py`:
  - `PatchDataset(torch.utils.data.Dataset)` with `__getitem__(idx) → field.select(sampler.anchors_[idx])`.
  - Distributed sampler that partitions anchors by rank (compatible with `DistributedSampler`).
  - Multi-worker shard: `worker_init_fn` divides anchors per worker.
- [ ] `geopatcher/jax/loaders.py`:
  - `PatchLoader` using `grain` (Google) or pure iterator + `jax.tree_util.tree_map`.
  - Prefetch via background thread.
- [ ] `[torch]` and `[jax]` extras.

### Testing plan

- [ ] `DataLoader` with `num_workers ∈ {0, 2, 4}` produces same anchors (seed-controlled).
- [ ] JAX loader: batch shape correct, anchor IDs round-trip through `unbatch`.
- [ ] Distributed: 2-rank `torch.distributed` test → anchors strictly partitioned, union = full set.
- [ ] Multi-epoch shuffle: same seed → same epoch order; different seed → different order.
